### DISCUSSION TITLE
All game logic in C#, Unreal used as UI only

### DISCUSSION THREAD

**Discussion by pavelkux** [nxrighthere/UnrealCLR#200] [General]
Hello.
This might be a really newbie question but I wanted to be sure how UnrealCLR integration works before we dive any further.

We have a game, which works entirely in C# (and old version of Mogre engine) and .NET. All the logic is contained within number of different projects referencing each other. The current engine is initiated within the UI project (standard WinForm project) and hence we have full control over every aspect of the game.

We would like to start working on a port and we are seriously considering Unreal and UnrealCLR. How much would we need to change?
I watched number of videos of UnrealCLR integration but those are rather very small projects and they also do not show how it would work with entire business layer projects.

To narrow it down, I am listing several questions but any information beside these is highly appreciated, too.
- Is UnrealCLR capable of including the whole solution structure, referencing between projects, etc?
- Are there any limitations in NET libraries usage?
- When working with UnrealCLR, can this be simply debugged in Visual Studio?
- How complicated is the build process when you change a line in .NET code of your project? What is involved in building?
- Would UnrealCLR allow to use such libraries like CSScript https://www.cs-script.net/
- Does UnrealCLR work on all platforms supported by Unreal itself?

If there are any links where these questions are already answered, please share them with me. I was searching for them but all I found were very broad and simple projects and answers.

Thank you!

**Comment by nxrighthere** [comment]
> Is UnrealCLR capable of including the whole solution structure, referencing between projects, etc?

The plugin itself cares only about the compiled assemblies that you feed to it, the structure of the project is up to you. You have to keep in mind that only one `UnrealEngine.Framework` assembly can be referenced in the project.

> Are there any limitations in NET libraries usage?

- If a particular library doesn't support [unloadability](https://github.com/nxrighthere/UnrealCLR/discussions/190#discussioncomment-1383229), then you have to take care of that.
- Managed libraries that are based on interoperability with multiple native libraries are supported [only partially](https://github.com/nxrighthere/UnrealCLR/discussions/152#discussioncomment-838836).

> When working with UnrealCLR, can this be simply debugged in Visual Studio?

Yes, absolutely.

> How complicated is the build process when you change a line in .NET code of your project? What is involved in building?

The building pipeline is entirely up to you. The plugin needs only assemblies of your project.

> Would UnrealCLR allow to use such libraries like CSScript https://www.cs-script.net/

You have to do some quick tests for this one. I don't know, unfortunately.

> Does UnrealCLR work on all platforms supported by Unreal itself?

Only desktop platforms are supported at the moment and primarily Windows. Unreal itself doesn't work well on, say, Linux, from my experience.

Speaking of your question using Unreal only as UI, I think the biggest problem is interoperability with managed code from widgets and so on. The built-in blueprints of the plugin are not flexible enough for that. Widgets API itself is [not exposed](https://github.com/nxrighthere/UnrealCLR/discussions/55) because it's tied to https://github.com/nxrighthere/UnrealCLR/issues/33.

On top of this, the plugin is in the preview stage. It has weak spots due to limitations on both sides .NET and Unreal Engine. In future, it will get better, but as a middleware plugin, I'm at the mercy of Microsoft and Epic.
